# 6. Forecasting Modelle

## 6.1 Vorbereitungen

Bevor wir Forecasting Modelle anwenden können müssen wir noch den Datensatz tageweise zusammenfassen

In [ ]:
# Löschen der Zeilen mit Werten über 20 in der Spalte 'butterzopf_mittel'
#data_prep = data_prep.loc[data_prep['butterzopf_mittel'] > 20, 'butterzopf_mittel'] = None

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 719678 entries, 0 to 719677
Data columns (total 20 columns):
 #   Column                       Non-Null Count   Dtype         
---  ------                       --------------   -----         
 0   Artikel                      719678 non-null  object        
 1   Menge                        719678 non-null  float64       
 2   Umsatz                       719678 non-null  float64       
 3   Bestellungen                 719678 non-null  float64       
 4   Niederschlag                 717181 non-null  float64       
 5   Temperatur Schnitt Tag       711720 non-null  float64       
 6   Niederschlag Summe Tag       719678 non-null  float64       
 7   Globalstrahlung Schnitt Tag  711720 non-null  float64       
 8   Uhrzeit                      719678 non-null  object        
 9   Stunde                       719678 non-null  int64         
 10  Datum                        719678 non-null  datetime64[ns]
 11  Wochentag                 

In [ ]:
# Verschiedene Werte in Spalte 'Ferienzeit TG'
data['Ferienzeit TG'].unique()

['Weihnachtsferien', NaN, 'Sportferien', 'Frühlingsferien', 'Pfingstferien', 'Sommerferien', 'Herbstferien']
Categories (6, object): ['Herbstferien' < 'Weihnachtsferien' < 'Sportferien' < 'Frühlingsferien' < 'Pfingstferien' < 'Sommerferien']

In [ ]:
# Add 'Keine Ferienzeit' as a new category to 'Ferienzeit TG' column
data['Ferienzeit TG'] = data['Ferienzeit TG'].cat.add_categories(['Keine Ferienzeit'])

# Fill missing values with the new category
data['Ferienzeit TG'] = data['Ferienzeit TG'].fillna('Keine Ferienzeit')

In [ ]:
# Droppen von Spalten, die nicht zur Regression beitragen
data_prep = data.drop(['Stückpreis', 'Wochentag', 'Umsatz', 'Niederschlag', 'Uhrzeit', 'Stunde', 'Feiertag Name'], axis = 1)
data_prep = pd.get_dummies(data_prep, columns = ['Ferienzeit TG', 'Geschmack', 'Vortag von'])
data_prep.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 719678 entries, 0 to 719677
Data columns (total 46 columns):
 #   Column                                           Non-Null Count   Dtype         
---  ------                                           --------------   -----         
 0   Artikel                                          719678 non-null  object        
 1   Menge                                            719678 non-null  float64       
 2   Bestellungen                                     719678 non-null  float64       
 3   Temperatur Schnitt Tag                           711720 non-null  float64       
 4   Niederschlag Summe Tag                           719678 non-null  float64       
 5   Globalstrahlung Schnitt Tag                      711720 non-null  float64       
 6   Datum                                            719678 non-null  datetime64[ns]
 7   Monat                                            719678 non-null  int64         
 8   Fastenzeit              

In [ ]:
# Entfernen von Missing Values außer in Dummies
data_prep = data_prep.dropna()
data_prep.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 711720 entries, 520 to 719677
Data columns (total 46 columns):
 #   Column                                           Non-Null Count   Dtype         
---  ------                                           --------------   -----         
 0   Artikel                                          711720 non-null  object        
 1   Menge                                            711720 non-null  float64       
 2   Bestellungen                                     711720 non-null  float64       
 3   Temperatur Schnitt Tag                           711720 non-null  float64       
 4   Niederschlag Summe Tag                           711720 non-null  float64       
 5   Globalstrahlung Schnitt Tag                      711720 non-null  float64       
 6   Datum                                            711720 non-null  datetime64[ns]
 7   Monat                                            711720 non-null  int64         
 8   Fastenzeit            

In [ ]:
data_prep.head()

Artikel  Menge  Bestellungen  Temperatur Schnitt Tag  \
520  Butterzopf mittel    1.0           0.0                5.245833   
521  Butterzopf mittel    1.0           0.0                5.245833   
522  Butterzopf mittel    1.0           0.0                5.245833   
523  Butterzopf mittel    1.0           0.0                5.245833   
524  Butterzopf mittel    1.0           0.0                5.245833   

     Niederschlag Summe Tag  Globalstrahlung Schnitt Tag      Datum  Monat  \
520                     1.8                    39.166667 2016-02-05      2   
521                     1.8                    39.166667 2016-02-05      2   
522                     1.8                    39.166667 2016-02-05      2   
523                     1.8                    39.166667 2016-02-05      2   
524                     1.8                    39.166667 2016-02-05      2   

     Fastenzeit  Ramadan  ...  Vortag von_Vortag Schweizer Vätertag  \
520           0        0  ...                                     0   
521           0        0  ...                                     0   
522           0        0  ...                                     0   
523           0        0  ...                                     0   
524           0        0  ...                                     0   

     Vortag von_Vortag Silvester  Vortag von_Vortag St. Nikolaus  \
520                            0                               0   
521                            0                               0   
522                            0                               0   
523                            0                               0   
524                            0                               0   

     Vortag von_Vortag Stephanstag  Vortag von_Vortag Tag der Arbeit  \
520                              0                                 0   
521                              0                                 0   
522                              0                                 0   
523                              0                                 0   
524                              0                                 0   

     Vortag von_Vortag Valentinstag  Vortag von_Vortag Vierter Advent  \
520                               0                                 0   
521                               0                                 0   
522                               0                                 0   
523                               0                                 0   
524                               0                                 0   

     Vortag von_Vortag Weihnachten  Vortag von_Vortag Zweiter Advent  \
520                              0                                 0   
521                              0                                 0   
522                              0                                 0   
523                              0                                 0   
524                              0                                 0   

     Vortag von_normaler Tag  
520                        1  
521                        1  
522                        1  
523                        1  
524                        1  

[5 rows x 46 columns]

In [ ]:
# One-Hot-Encoding für die Spalte "Artikel"
#data_prep = pd. get_dummies (data_prep, columns = [ 'Artikel' ])

# Gruppieren der Daten nach Datum und berechnen der Summe der Menge pro Artikel
sum_cols = ['Menge', 'Bestellungen']

mean_cols = ['Temperatur Schnitt Tag','Niederschlag Summe Tag', 
             'Globalstrahlung Schnitt Tag', 'Fastenzeit','Ramadan']

for col in data_prep.columns:
    if col.startswith(('Ferienzeit', 'Geschmack', 'Feiertag', 'Vortag von')):
        mean_cols.append(col)

agg_dict = {col: 'sum' for col in sum_cols}
agg_dict.update({col: 'mean' for col in mean_cols})

data_prep = data_prep.groupby(['Datum', 'Artikel']).agg(agg_dict).reset_index()

# Ersetzen der Leerzeichen in Artikel mit Unterstrich
data_prep['Artikel'] = data_prep['Artikel'].str.replace(' ', '_')
data_prep = data_prep.drop('Datum', axis = 1)
data_prep.head()


Artikel  Menge  Bestellungen  Temperatur Schnitt Tag  \
0           BT_Nussgipfel   14.0           0.0                3.890909   
1           Buttergipfeli  131.0           0.0                3.890909   
2          Crèmeschnitten   28.0           0.0                3.890909   
3     Hefe-Vanille-Weggli   13.0           0.0                3.890909   
4  Il_pollo_forte_gefüllt   20.0           0.0                3.890909   

   Niederschlag Summe Tag  Globalstrahlung Schnitt Tag  Fastenzeit  Ramadan  \
0                     1.6                    12.636364         0.0      0.0   
1                     1.6                    12.636364         0.0      0.0   
2                     1.6                    12.636364         0.0      0.0   
3                     1.6                    12.636364         0.0      0.0   
4                     1.6                    12.636364         0.0      0.0   

   Ferienzeit TG_Herbstferien  Ferienzeit TG_Weihnachtsferien  ...  \
0                         0.0                             0.0  ...   
1                         0.0                             0.0  ...   
2                         0.0                             0.0  ...   
3                         0.0                             0.0  ...   
4                         0.0                             0.0  ...   

   Vortag von_Vortag Schweizer Vätertag  Vortag von_Vortag Silvester  \
0                                   0.0                          0.0   
1                                   0.0                          0.0   
2                                   0.0                          0.0   
3                                   0.0                          0.0   
4                                   0.0                          0.0   

   Vortag von_Vortag St. Nikolaus  Vortag von_Vortag Stephanstag  \
0                             0.0                            0.0   
1                             0.0                            0.0   
2                             0.0                            0.0   
3                             0.0                            0.0   
4                             0.0                            0.0   

   Vortag von_Vortag Tag der Arbeit  Vortag von_Vortag Valentinstag  \
0                               0.0                             0.0   
1                               0.0                             0.0   
2                               0.0                             0.0   
3                               0.0                             0.0   
4                               0.0                             0.0   

   Vortag von_Vortag Vierter Advent  Vortag von_Vortag Weihnachten  \
0                               0.0                            0.0   
1                               0.0                            0.0   
2                               0.0                            0.0   
3                               0.0                            0.0   
4                               0.0                            0.0   

   Vortag von_Vortag Zweiter Advent  Vortag von_normaler Tag  
0                               0.0                      1.0  
1                               0.0                      1.0  
2                               0.0                      1.0  
3                               0.0                      1.0  
4                               0.0                      1.0  

[5 rows x 44 columns]

In [ ]:
# Splitten des Datensets in Datasets für die einzelnen Artikel mit einer for-schleife

dataframes = [data_prep[data_prep['Artikel'] == i] for i in data_prep['Artikel'].unique()]

In [ ]:
data_prep.reset_index(drop=True, inplace=True)
data_prep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27856 entries, 0 to 27855
Data columns (total 44 columns):
 #   Column                                           Non-Null Count  Dtype  
---  ------                                           --------------  -----  
 0   Artikel                                          27856 non-null  object 
 1   Menge                                            27856 non-null  float64
 2   Bestellungen                                     27856 non-null  float64
 3   Temperatur Schnitt Tag                           27856 non-null  float64
 4   Niederschlag Summe Tag                           27856 non-null  float64
 5   Globalstrahlung Schnitt Tag                      27856 non-null  float64
 6   Fastenzeit                                       27856 non-null  float64
 7   Ramadan                                          27856 non-null  float64
 8   Ferienzeit TG_Herbstferien                       27856 non-null  float64
 9   Ferienzeit TG_Weihnachtsferi

In [ ]:
data_prep.head()

Artikel  Menge  Bestellungen  Temperatur Schnitt Tag  \
0           BT_Nussgipfel   14.0           0.0                3.890909   
1           Buttergipfeli  131.0           0.0                3.890909   
2          Crèmeschnitten   28.0           0.0                3.890909   
3     Hefe-Vanille-Weggli   13.0           0.0                3.890909   
4  Il_pollo_forte_gefüllt   20.0           0.0                3.890909   

   Niederschlag Summe Tag  Globalstrahlung Schnitt Tag  Fastenzeit  Ramadan  \
0                     1.6                    12.636364         0.0      0.0   
1                     1.6                    12.636364         0.0      0.0   
2                     1.6                    12.636364         0.0      0.0   
3                     1.6                    12.636364         0.0      0.0   
4                     1.6                    12.636364         0.0      0.0   

   Ferienzeit TG_Herbstferien  Ferienzeit TG_Weihnachtsferien  ...  \
0                         0.0                             0.0  ...   
1                         0.0                             0.0  ...   
2                         0.0                             0.0  ...   
3                         0.0                             0.0  ...   
4                         0.0                             0.0  ...   

   Vortag von_Vortag Schweizer Vätertag  Vortag von_Vortag Silvester  \
0                                   0.0                          0.0   
1                                   0.0                          0.0   
2                                   0.0                          0.0   
3                                   0.0                          0.0   
4                                   0.0                          0.0   

   Vortag von_Vortag St. Nikolaus  Vortag von_Vortag Stephanstag  \
0                             0.0                            0.0   
1                             0.0                            0.0   
2                             0.0                            0.0   
3                             0.0                            0.0   
4                             0.0                            0.0   

   Vortag von_Vortag Tag der Arbeit  Vortag von_Vortag Valentinstag  \
0                               0.0                             0.0   
1                               0.0                             0.0   
2                               0.0                             0.0   
3                               0.0                             0.0   
4                               0.0                             0.0   

   Vortag von_Vortag Vierter Advent  Vortag von_Vortag Weihnachten  \
0                               0.0                            0.0   
1                               0.0                            0.0   
2                               0.0                            0.0   
3                               0.0                            0.0   
4                               0.0                            0.0   

   Vortag von_Vortag Zweiter Advent  Vortag von_normaler Tag  
0                               0.0                      1.0  
1                               0.0                      1.0  
2                               0.0                      1.0  
3                               0.0                      1.0  
4                               0.0                      1.0  

[5 rows x 44 columns]

In [ ]:
pre_optimized = data_prep.copy()

### Optimierung

In [ ]:
# Für Butterzopf_mittel: Berechnung der minimalen und maximalen Menge, sowie der Gesamtsumme der Menge
butterzopf_mittel = pre_optimized[data_prep['Artikel'] == 'Butterzopf_mittel']
print(butterzopf_mittel['Menge'].describe())
print('Summe: ', butterzopf_mittel['Menge'].sum())
print('Median: ', butterzopf_mittel['Menge'].median())

count    1975.000000
mean       31.604051
std        32.702989
min         1.000000
25%         7.000000
50%        15.000000
75%        57.000000
max       312.000000
Name: Menge, dtype: float64
Summe:  62418.0
Median:  15.0


Beim Butterzopf ist auffällig, dass bei einer relativ geringen Menge (16% von Verkaufsmenge Buttergipfel) eine hohe Variation in den gekauften Mengen besteht. Das erkennt man besonders daran, dass der Durchschnitt bei 31.6 liegt, während der Median nur bei 15 liegt.

Um dem entgegenzuwirken entfernen wir Werte größer 20, da wir davon ausgehen, dass solch große Mengen nicht ohne Voranküdigung gekauft werden. Und wenn der Kauf vorbestellt ist, dann ist er auch nicht mehr für unsere Prediciton relevant.

In [ ]:
# For schleife, die durch alle Artikel iteriert und die Zeile löscht, wenn der Artikel 'butteropf Mittel' größer als 20 ist
#for i in range(len(data_prep)):
#    if data_prep['Artikel'][i] == 'Butterzopf_mittel' and data_prep['Menge'][i] > 20:
#        data_prep.drop(i, inplace = True)

#data_prep.info()

## 6.2 Lineare Regression

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_percentage_error
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression

In [ ]:
def LinReg_train_and_evaluate_product_model(data, product_id):
    # Filter the data for the specific product
    product_data = data[data['Artikel'] == product_id]
    
    # Split the data into features (X) and target (y)
    X = product_data.drop(['Menge', 'Artikel'], axis=1)
    y = product_data['Menge']
    
    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # Define the XGBoost model
    model = LinearRegression()
    
    # Train the model on the training data
    model.fit(X_train, y_train)
    
    # Make predictions on the testing data
    y_pred = model.predict(X_test)
    
    # Calculate the mean squared error of the model
    MAPE = mean_absolute_percentage_error(y_test, y_pred)
    R2 = model.score(X_test, y_test)
    RMSE = np.sqrt(mean_squared_error(y_test, y_pred))
    
    return MAPE, R2, RMSE


In [ ]:
LinReg_train_and_evaluate_product_model(data_prep, 'Buttergipfeli')

(0.2310273469899282, 0.4887705386642286, 56.39523916614759)

In [ ]:
linReg_Artikel = []  # Create an empty list to store the models

metric_names = ['MAPE', 'R2', 'RMSE']  # Add more metric names if there are more than 3 returned values

for product in data_prep['Artikel'].unique():
    model_metrics = LinReg_train_and_evaluate_product_model(data_prep, product)
    model_metrics_dict = {'Type': 'LinReg', 'product': product, **dict(zip(metric_names, model_metrics))}
    linReg_Artikel.append(model_metrics_dict)

# Visualize the results
linReg_Artikel = pd.DataFrame(linReg_Artikel)
linReg_Artikel

Type                 product          MAPE            R2          RMSE
0   LinReg           BT_Nussgipfel      0.262727  4.720467e-01  1.030151e+01
1   LinReg           Buttergipfeli      0.231027  4.887705e-01  5.639524e+01
2   LinReg          Crèmeschnitten      0.311112  7.212602e-01  1.154892e+01
3   LinReg     Hefe-Vanille-Weggli      0.256981  7.771647e-01  4.370407e+00
4   LinReg  Il_pollo_forte_gefüllt      0.159996  6.714444e-01  3.084070e+00
5   LinReg            Laugenbrötli      0.226687  5.732227e-01  1.163861e+01
6   LinReg      Laugengi.mit_SoBlu      0.203696  6.503116e-01  1.974626e+01
7   LinReg              Maisgipfel      0.240854  6.727269e-01  1.885227e+01
8   LinReg             Quarkzöpfli      0.192755  6.964544e-01  4.762058e+00
9   LinReg           Schoggibrötli      0.422063  8.305662e-01  6.821888e+00
10  LinReg         St._Gallerbürli  35272.896373 -1.750166e+12  2.024046e+07
11  LinReg           Weggli_normal      0.257100  3.688393e-01  1.267601e+01
12  LinReg       Butterzopf_mittel      0.942959  7.645928e-01  1.515364e+01
13  LinReg           Erdbeertörtli      0.260792  4.938781e-01  1.124031e+01

In [ ]:
# Extrahieren von MAPE aus dem Dataframe
LinReg_Artikel_MAPE = linReg_Artikel[['product', 'MAPE']]
LinReg_Artikel_MAPE

product          MAPE
0            BT_Nussgipfel      0.262727
1            Buttergipfeli      0.231027
2           Crèmeschnitten      0.311112
3      Hefe-Vanille-Weggli      0.256981
4   Il_pollo_forte_gefüllt      0.159996
5             Laugenbrötli      0.226687
6       Laugengi.mit_SoBlu      0.203696
7               Maisgipfel      0.240854
8              Quarkzöpfli      0.192755
9            Schoggibrötli      0.422063
10         St._Gallerbürli  35272.896373
11           Weggli_normal      0.257100
12       Butterzopf_mittel      0.942959
13           Erdbeertörtli      0.260792

In [ ]:
# Visualisieren der Ergebnisse
import matplotlib.pyplot as plt
import seaborn as sns


## 6.3 XGBoost Regressor

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_percentage_error
from sklearn.model_selection import train_test_split
import xgboost as xgb

In [ ]:
buttergipfel = data_prep[data_prep['Artikel'] == 'Buttergipfeli']
buttergipfel.head()

# Splitten des Datensets in Trainings- und Testdaten
X = buttergipfel.drop(['Menge', 'Artikel'], axis=1)
y = buttergipfel['Menge']

X_buttergipfel_train, X_buttergipfel_test, y_buttergipfel_train, y_buttergipfel_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Erstellen des Modells
model_Buttergipfeli_xgb = model.XGBRegressor(objective ='reg:squarederror')

# Trainieren des Modells
model_Buttergipfeli_xgb.fit(X_buttergipfel_train, y_buttergipfel_train)

# Vorhersagen mit dem Modell
y_buttergipfel_pred = xgb.predict(X_buttergipfel_test)

# Berechnen des MAPE
mape = mean_absolute_percentage_error(y_buttergipfel_test, y_buttergipfel_pred)
print('MAPE: %.3f' % mape)

NameError: name 'model' is not defined

In [ ]:
def XGB_train_and_evaluate_product_model(data, product_id):
    # Filter the data for the specific product
    product_data = data[data['Artikel'] == product_id]
    
    # Split the data into features (X) and target (y)
    X = product_data.drop(['Menge', 'Artikel'], axis=1)
    y = product_data['Menge']
    
    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    
    # Define the XGBoost model
    model = xgb.XGBRegressor(objective='reg:squarederror', n_jobs=-1)
    
    # Train the model on the training data
    model.fit(X_train, y_train)
    
    # Make predictions on the testing data
    y_pred = model.predict(X_test)
    
    # Calculate the mean squared error of the model
    MAPE = mean_absolute_percentage_error(y_test, y_pred)
    R2 = model.score(X_test, y_test)
    RMSE = np.sqrt(mean_squared_error(y_test, y_pred))
    
    return MAPE, R2, RMSE


In [ ]:
xgb_Artikel = []  # Create an empty list to store the models

metric_names = ['MAPE', 'R2', 'RMSE']  # Add more metric names if there are more than 3 returned values

for product in data_prep['Artikel'].unique():
    model_metrics = XGB_train_and_evaluate_product_model(data_prep, product)
    model_metrics_dict = {'Type': 'XGBoost', 'product': product, **dict(zip(metric_names, model_metrics))}
    xgb_Artikel.append(model_metrics_dict)

# Visualize the results
xgb_Artikel = pd.DataFrame(xgb_Artikel)
xgb_Artikel


In [ ]:
# Extrahieren von MAPE aus dem Dataframe
xgb_Artikel_MAPE = xgb_Artikel[['product', 'MAPE']]
xgb_Artikel_MAPE

## 6.5 Vergleich der Modelle

In [ ]:
# Verbinden der beiden Dataframes
Artikel_MAPE = pd.merge(LinReg_Artikel_MAPE, xgb_Artikel_MAPE, on='product')

# Umbenennen der Spalten
Artikel_MAPE.columns = ['Artikel', 'Lineare Regression', 'XGB']
Artikel_MAPE

In [ ]:
# Berechnen des MAPE-Differenz
Artikel_MAPE['MAPE Diff XGB'] = Artikel_MAPE['XGB'] - Artikel_MAPE['Lineare Regression']
Artikel_MAPE.round(4)

In [ ]:
bar_width = 0.35

# Positions of the bars on the x-axis
ind = np.arange(len(Artikel_MAPE['Artikel']))

# Figure size
plt.figure(figsize=(14, 6))

# Create bars
plt.bar(ind, Artikel_MAPE['Lineare Regression'], width=bar_width, label='Lineare Regression')
plt.bar(ind + bar_width, Artikel_MAPE['XGB'], width=bar_width, label='XGB')

# Set x-axis labels and ticks
plt.xticks(ind + bar_width / 2, Artikel_MAPE['Artikel'], rotation=45)

# Set y-axis label
plt.ylabel('Values')

# Set plot title
plt.title('Linear Regression vs. XGB Values for Each Artikel')

# Add legend
plt.legend(loc='best')

# Show plot
plt.tight_layout()
plt.show()

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

def custom_palette(values):
    palette = []
    for value in values:
        if value >= 0:
            palette.append('red')
        else:
            palette.append('green')
    return palette

plt.figure(figsize=(10, 5))

# Extract MAPE Diff XGB values for the custom_palette function
mape_diff_values = Artikel_MAPE['MAPE Diff XGB'].values

# Use the custom palette with the sns.barplot function
sns.barplot(x='Artikel', y='MAPE Diff XGB', data=Artikel_MAPE, palette=custom_palette(mape_diff_values))

plt.xticks(rotation=90)
plt.show()


In [ ]:
def plot_variable_importance(model_Buttergipfeli_xgb, X_buttergipfel_train):
    import matplotlib.pyplot as plt
    from pandas import DataFrame
    imp=DataFrame({"imp":model_Buttergipfeli_xgb.feature_importances_, "names":X_buttergipfel_train.columns}).sort_values("imp", ascending=True)
    fig, ax = plt.subplots(figsize=(imp.shape[0]/6,imp.shape[0]/5), dpi=300) 
    ax.barh(imp["names"],imp["imp"], color="green") 
    ax.set_xlabel('\nVariable Importance')
    ax.set_ylabel('Features\n') 
    ax.set_title('Variable Importance Plot\n') 
    plt.show() 
plot_variable_importance(model_Buttergipfeli_xgb, X_buttergipfel_train)

# Optimierung des Datensatzes

## Test-Optimierung des XGBoost Modells mit Cross Validation und Grid Search für Buttergipfeli

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
import xgboost as xgb
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, mean_absolute_percentage_error

# Initialisieren des Modells
xgb = xgb.XGBRegressor()

param_grid = {
    'learning_rate': [0.01, 0.05, 0.1],
    'max_depth': [3, 5, 7],
    'n_estimators': [100, 300],
    'subsample': [0.8, 0.9],
    'colsample_bytree': [0.8, 0.9],
}

# Erschaffen einer GridSearchCV object
xgb_CV_Buttergipfel = GridSearchCV(xgb, param_grid, scoring='r2', cv=5, n_jobs=-1, verbose=1)
xgb_CV_Buttergipfel.fit(X_buttergipfel_train, y_buttergipfel_train)

# Bestes Modell speichern
best_xgb = xgb_CV_Buttergipfel.best_estimator_

# KPIs berechnen
best_model_R2 = best_xgb.score(X_buttergipfel_train, y_buttergipfel_train)
y_pred_xgb = best_xgb.predict(X_buttergipfel_test)
best_model_RMSE = mean_squared_error(y_buttergipfel_test, y_pred_xgb, squared=False)
best_model_MAPE = mean_absolute_percentage_error(y_buttergipfel_test, y_pred_xgb)

print("Best Model R2:", best_model_R2)
print("Best Model RMSE:", best_model_RMSE)
print("Best Model MAPE:", best_model_MAPE)

In [ ]:
param_grid = {
    'learning_rate': [0.01, 0.05, 0.1],
    'max_depth': [3, 5, 7],
    'n_estimators': [100, 300],
    'subsample': [0.8, 0.9],
    'colsample_bytree': [0.8, 0.9],
}

# Erstellen des GridSearchCV-Objekts
xgb_CV_Buttergipfel = GridSearchCV(xgb, param_grid, scoring = 'mape', cv = 5, n_jobs = -1, verbose = 1)
xgb_CV_Buttergipfel.fit(X_buttergipfel_train, y_buttergipfel_train) # fitting des Modells

# Ausgabe der besten Parameter
best_xgb = xgb_CV_Buttergipfel.best_estimator_

# Berechnen der KPI
best_model_R2 = best_xgb.score(X_buttergipfel_train, y_buttergipfel_train)
y_pred_xgb = best_xgb.predict(X_buttergipfel_test)
best_model_RMSE = mean_squared_error(y_buttergipfel_test, y_pred_xgb, squared=False)
best_model_MAPE = mean_absolute_percentage_error(y_buttergipfel_test, y_pred_xgb)

print("Best Model R2:", best_model_R2)
print("Best Model RMSE:", best_model_RMSE)
print("Best Model MAPE:", best_model_MAPE)

## Test zum Anpassen von Extremwerten in der Menge

Vor allem der 'Butterzopf_mittel' sticht in der Analyse heraus. Er hat einen unglaublich hohen MAPE mit 62% Abweichung aber auch ein hohes R2 (0.84). Deshalb nehmen wir diesen Artikel als Startpunkt auf der Suche nach besserem Verständnis unseres Modells und den Eigenheiten des Datensatzes

In [ ]:
# Boxplot aller Artikel
plt.figure(figsize=(10, 5))
sns.boxplot(x='Artikel', y='Menge', data=data_prep)
plt.xticks(rotation=90)
plt.show() 

In [ ]:
# Für Butterzopf_mittel: Berechnung der minimalen und maximalen Menge, sowie der Gesamtsumme der Menge
butterzopf_mittel = pre_optimized[data_prep['Artikel'] == 'Butterzopf_mittel']
print(butterzopf_mittel['Menge'].describe())
print('Summe: ', butterzopf_mittel['Menge'].sum())
print('Median: ', butterzopf_mittel['Menge'].median())


Beim Butterzopf ist auffällig, dass bei einer relativ geringen Menge (16% von Verkaufsmenge Buttergipfel) eine hohe Variation in den gekauften Mengen besteht. Das erkennt man besonders daran, dass der Durchschnitt bei 31.6 liegt, während der Median nur bei 15 liegt.

Um dem entgegenzuwirken entfernen wir Werte größer 20, da wir davon ausgehen, dass solch große Mengen nicht ohne Voranküdigung gekauft werden. Und wenn der Kauf vorbestellt ist, dann ist er auch nicht mehr für unsere Prediciton relevant.

In [ ]:

butterzopf_mittel = butterzopf_mittel[butterzopf_mittel['Menge'] < 20]
butterzopf_test = XGB_train_and_evaluate_product_model(butterzopf_mittel, 'Butterzopf_mittel')
print(butterzopf_test)


Mit sinkendem Maximalen Wert, ist die der MAPE ebenfalls. Deshalb wenden wir diese Cleaning Maßnahme nun auch im richtigen Modell an

In [ ]:
# Für Schoggibrötli: Berechnung der minimalen und maximalen Menge, sowie der Gesamtsumme der Menge
Schoggibrötli = data_prep[data_prep['Artikel'] == 'Schoggibrötli']
print(Schoggibrötli['Menge'].describe())
print('Summe: ', Schoggibrötli['Menge'].sum())
print('Median: ', Schoggibrötli['Menge'].median())

In [ ]:
# Für Butterzopf_mittel: Berechnung der minimalen und maximalen Menge, sowie der Gesamtsumme der Menge
buttergipfeli = data_prep[data_prep['Artikel'] == 'Buttergipfeli']
print(buttergipfeli['Menge'].describe())
print('Summe: ', buttergipfeli['Menge'].sum())
print('Median: ', buttergipfeli['Menge'].median())

Im Vergleich mit dem Buttergipfel zeigt sich, dass 

# Ablage